_posts/2020-05-22-validate-spring-boot-configuration-parameters-at-startup.md

```diff
@@ -0,0 +1,225 @@
+***************************
+
+Description:
+
+Binding to target org.springframework.boot.context.properties.bind.BindException: Failed to bind properties under 'app.properties' to io.reflectoring.validation.AppProperties failed:
+
+    Property: app.properties.name
+    Value:
+    Origin: class path resource [application-validation.properties]:1:21
+    Reason: must not be empty
+
+
+Action:
+
+Update your application's configuration
+```
+
+<div class="notice warning">
+  <h4>Bean Validation API Dependency</h4>
+  In order to make bean validation work, we must have <code>javax.validation.validation-api</code> dependency in our classpath. For Spring Boot, take a look at <a href="https://reflectoring.io/bean-validation-with-spring-boot/#setting-up-validation">setting up bean validation</a> section.
+</div>
+
+Additionally, we can also define some default values that are set when our `AppProperties` class is initialized:
+```java
+
+private Boolean sendReportEmails = Boolean.FALSE;
+
+private ReportType reportType = ReportType.HTML;
+
+```
+Therefore, even if we don't define any values for the properties `send-report-emails` and `report-type`, we can still get the default values `Boolean.FALSE` and `ReportType.HTML` respectively.
+
+### Nested Configuration Objects
+For some properties, it would be meaningful to bind them into a nested object.
+
+So, let's create `ReportProperties` to group the properties related to `report` key:
+
+```java
+class ReportProperties {
+
+  private Boolean sendEmails = Boolean.FALSE;
+
+  private ReportType type = ReportType.HTML;
+
+  @Min(value = 7)
+  @Max(value = 30)
+  private Integer intervalInDays;
+
+  @Email
+  private String emailAddress;
+
+  // ...
+}
+```
+
+Next, we refactor our `AppProperties` to include our nested object `ReportProperties`:
+
+```java
+@Validated
+@ConfigurationProperties(prefix="app.properties")
+class AppProperties {
+
+  @NotEmpty
+  private String name;
+
+  @Valid
+  private ReportProperties report;
+
+  // ...
+}
+```
+
+Besides, we should also pay attention to put `@Valid` annotation on our nested class field `report`.
+
+Thus, Spring can validate the inner properties included in nested objects.
+
+### Using Bean Methods
+Another way of triggering validation, we can use the Spring bean methods. This is particularly useful when we want to **bind properties to components defined in third-party libraries or maintained in separate jar files**:
+```java
+@Bean
+@Validated
+@ConfigurationProperties(prefix = "app.third-party.properties")
+public ThirdPartyComponentProperties thirdPartyComponentProperties() {
+  return ThirdPartyComponentProperties();
+}
+```
+
+### Using a Custom `Validator`
+Even though bean validation provides a declarative approach to validate our objects in a reusable way, sometimes we need more to customize our validation logic in purpose.
+
+In that case, **Spring has an independent `Validator` mechanism to allow dynamic input validation**.
```

Comment:
That `Validator` doesn't have anything to do with Bean Validation, right? Could you add that in a sentence?

Reply:
I've added a warning field about that.